### PR TITLE
Add support for crc check failure handling

### LINKED ISSUE #306: later messages are not parsed correctly when parsing of one previous message failed
#### Problem

In current message parser, when there is a format failure, the parser simply return null and asks for more recving data for trying again later. This blocks the later messages when the current message is mal-formatted instead of the data the incomplete. 
#### Solution

need to change the base class of message parser so it can close the rpc session to allow for further messages come in using a new session (and parser).

### PR DESCRIPTION
**Pull request by shengofsun**
1. when receiving message data's crc is not correct, call on_failure. partially resolve #306 
2. add test for handling crc's failure
3. refactor the test cases in core/tests, currently a "gtest.fiter" file is added, which describes the gtest cases for each "config-test-xxx.ini".

### CODE REVIEW

File: src/core/tools/hpc/hpc_network_provider.bsd.cpp [hunk 1]
```diff
@@ -241,11 +241,17 @@ namespace dsn
                 {
                     message_ex* msg = _parser->get_message_on_receive(sz, read_next);
 
-                    while (msg != nullptr)
+                    while (read_next != -1 && msg != nullptr)
```
Comment (imzhenyu) [inline src/core/tools/hpc/hpc_network_provider.bsd.cpp]: msg == nullptr when read_next == -1, right?

Comment (shengofsun) [inline src/core/tools/hpc/hpc_network_provider.bsd.cpp]: That's right.  Comments should be add in the base class of "message_parser", and this is my fault.

File: src/core/core/message_parser.cpp [hunk 2]
```diff
@@ -135,21 +135,43 @@ namespace dsn {
...
         {            
+            if ( !_header_checked && !message_ex::is_right_header((char*)_read_buffer.data()) )
+            {
+                derror("receive message header check failed for message");
+                read_next = -1;
+                //we just remove all data received when encounter checksum fail
+                _read_buffer_occupied = 0;
+                return nullptr;
+
+                _header_checked = true;
```
Comment (qinzuoyan) [inline src/core/core/message_parser.cpp]: set after return ?

**Comment by shengofsun** [comment]
Besides, I think the crc check feature should be a configuration option for rpc_channel, or for rpc_message, but not for the rpc_engine. What do you think? @imzhenyu @qinzuoyan

**Comment by imzhenyu** [comment]
@shengofsun I agree. We can make it an option inside task-spec, which allows configuration for each kind of rpc request/response.

**Comment by imzhenyu** [comment]
There are some rest results file, please remove.

**Comment by shengofsun** [comment]
@imzhenyu very sorry for this, have removed now.

File: src/core/tools/hpc/hpc_network_provider.bsd.cpp (hunk 1 above)
Comment (imzhenyu) [inline src/core/tools/hpc/hpc_network_provider.bsd.cpp]: I mean the condition does not need to be changed, right? As `msg != nullptr` always implies `read_next != -1`

**Comment by imzhenyu** [comment]
@shengofsun  No problem, that happens all the time:)

File: src/core/tools/hpc/hpc_network_provider.bsd.cpp (hunk 1 above)
Comment (imzhenyu) [inline src/core/tools/hpc/hpc_network_provider.bsd.cpp]: @shengofsun ping in case you did not see this ...

Comment (qinzuoyan) [inline src/core/tools/hpc/hpc_network_provider.bsd.cpp]: yes, you are right, I'll fix it.

**Pull request merged**